[ESP32-C3 / -S3] Add support for ESP32-C3 & ESP32-S3

TD-er opened the pull request:
#fixes: #4614
#fixes: #3863

ToDo ESP32-C3:
- [x] Implement pin definitions
- [x] Check settings struct for boot states
- [x] Implement hardware specifics (e.g. USB behavior, SPI busses)
- [x] Add proper defaults for SPI busses etc.
- [x] Fix determine free stack
- [x] Document
- [x] Test

ToDo ESP32-S3:
- [x] Implement pin definitions
- [x] Check settings struct for boot states
- [x] Implement hardware specifics (e.g. USB behavior, SPI busses)
- [x] Add proper defaults for SPI busses etc.
- [x] Document
- [x] Test


N.B. support for CDC (console via native USB) will be added later in a separate PR.

TD-er committed:
[ESP32-C3 / -S3] Add support for ESP32-C3 & ESP32-S3

TD-er committed:
[ESP32-C3] Disable P097 ESP32Touch for ESP32-C3

TD-er committed:
[GPIO info] Define flash pins for ESP32-C3 and ESP32-S3

iz8mbw commented:
Great enhancement!!
Thanks.

TD-er committed:
[GPIO info] Add hardware/GPIO specifics for ESP32-C3/S3

tonhuisman commented on lib/ESPEasySerial/ESPeasySerial.h:
```diff
@@ -41,19 +41,32 @@
+# elif defined(ESP32_CLASSIC)
+  #define HAS_SERIAL2 1
+# else
+  static_assert(false, "Implement processor architecture");
```
You might want to hint at ESPEasySerial in the message, or should the filename be explicit enough?

TD-er replied:
Not sure what you mean here.

tonhuisman replied:
The message doesn't include 'ESPEasySerial', but the assertion will probably include the filename, so I'm not sure an explict mention is needed.

TD-er replied:
Already worked around it by defining this in only 1 place using the defines of the ESP32-xx HAL defines :)

tonhuisman commented on boards/esp32-cam.json:
```diff
@@ -5,7 +5,7 @@
     },
     "core": "esp32",
-    "extra_flags": "-DARDUINO_ESP32_DEV -DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw -DARDUINO_USB_CDC_ON_BOOT=0 -DESP32_4M",
+    "extra_flags": "-DARDUINO_ESP32_DEV -DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw -DARDUINO_USB_CDC_ON_BOOT=0 -DESP32_4M -DESP32_CLASIC",
```
is that meant to be `ESP32_CLASSIC` with a double `S` ? (same in the other .json files)

TD-er replied:
Already fixed that locally as it was indeed a typo and I was quite confused to why I needed to define it in the platformio ini files too :)

TD-er committed:
[GPIO Boot States] Fix bug storing ESP32-S2 boot states + add ESP32-S3

TD-er committed:
[Build] Fix missing include

TD-er committed:
[Build] Fix build for ESP8266 missing a define macro

TD-er committed:
[NeoPixelBus] Update NeoPixelBus library to 2.7.5

TD-er committed:
[ESP32-S3] Fix board definitions + build errors on ESP32-S3

TD-er committed:
[ESP32-S3] Add 16M max_ESP32s3_16M8M_LittleFS build

TD-er commented:
Yay!
The 16MB LilyGo T-Dongle S3 is working:

![image](https://user-images.githubusercontent.com/3751318/235813616-0720f5af-defd-49d8-8346-1e8d5f36e0e4.png)

```
CLK: GPIO-5
MISO: GPIO-13
MOSI: GPIO-3
CS: GPIO-4
DC: GPIO-2
RES: GPIO-1
Btn: GPIO-0
Bckl: GPIO-38
```

TD-er committed:
[ESP32-C3] Fix computing free stack on ESP32-C3

TD-er commented:
Would appreciate if somebody could also test this PR on ESP32-C3 / S3 hardware as it has now been tested on a whoppin' 2 devices.
However it seems to be working quite well at the moment.
It is only lacking documentation and maybe some proper defaults for stuff like SPI pins.

Support for CDC (console via native USB) will be added later in another PR.

chemmex commented:
Testing on a ESP32C3_2M using @Jason2866 board template, so far have built and uploaded without problems. Will test SPI and other peripherals later today.

TD-er commented:
For S3 boards with PSRAM, I noticed I needed to use `qio_qspi` as memory type.
Otherwise the PSRAM would not be detected.

Thus for example using this on the platformIO envs:
```
board                     = esp32s3-qio_opi
```

Not sure if something similar is needed for C3 boards.  (nor if there are even C3 boards with PSRAM)

tonhuisman commented:
> Testing on a ESP32C3_2M...

Is that a C3 board with 2MB flash, or using a 2MB filesystem (though that's unlikely)? As with 2MB flash you will need an adjusted partition table, AFAICS.

chemmex commented:
> Is that a C3 board with 2MB flash, or using a 2MB filesystem (though that's unlikely)? As with 2MB flash you will need an adjusted partition table, AFAICS.

It is AI-Thinker ESP-C3-12F module with 2MB external SPI Flash. I've borrowed @Jason2866 's partition file, namely esp32_2M_partition_app1792k_no_ota_spiffs192k.csv

Jason2866 commented:
> > Testing on a ESP32C3_2M...
> 
> Is that a C3 board with 2MB flash, or using a 2MB filesystem (though that's unlikely)? As with 2MB flash you will need an adjusted partition table, AFAICS.

Yep, and NOT recommended since no OTA possible. Just did for the poor people bought the 2M module by accident

```
# Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     0x9000,  0x5000,
otadata,  data, ota,     0xe000,  0x2000,
app0,     app,  ota_0,   0x10000, 0x1C0000,
spiffs,   data, spiffs,  0x1D0000,0x30000,
```

chemmex commented:
> Yep, and NOT recommended since no OTA possible. Just did for the poor people bought the 2M module by accident

I would dream about having a squeezed OTA build for it as I've got a bunch of these after above mentioned accidental purchase

Jason2866 commented:
> > Yep, and NOT recommended since no OTA possible. Just did for the poor people bought the 2M module by accident
> 
> I would dream about having a squeezed OTA build for it as I've got a bunch of these after above mentioned accidental purchase

OTA NO way. The needed flash space for a minimal version doing the OTA process is WAAAY to big.
Even the OTA process of the ESP32x would support `zip` firmwares

Have you removed the shield metal from one of the modules? Maybe it is not a embedded flash? > Flash swap...

chemmex commented:
> ToDo ESP32-C3:
> * [ ]  Add proper defaults for SPI busses etc.

I think that old VSPI and HSPI presets should be deleted and only user defined left

Jason2866 commented:
> > ToDo ESP32-C3:
> > 
> > * [ ]   Add proper defaults for SPI busses etc.
> 
> I think that old VSPI and HSPI presets should be deleted and only user defined left

This will brake nearly all 3rd party libs used, or EspEasy is very different in handling this stuff than Tasmota.

chemmex commented:
> Have you removed the shield metal from one of the modules? Maybe it is not a embedded flash? > Flash swap...

It's definitely an external flash version, I tried to swap it with 4MB chip, but no luck so far, it doesn't boot in serial download mode. Will try again

Jason2866 commented:
> > Have you removed the shield metal from one of the modules? Maybe it is not a embedded flash? > Flash swap...
> 
> It's definitely an external flash version, I tried to swap it with 4MB chip, but no luck so far, it doesn't boot in serial download mode. Will try again

Thats for sure a solder issue.

TD-er committed:
[Docs] Document pins to use for ESP32-S2/S3/C3

TD-er committed:
[PSRAM] Fix PSRAM detection on ESP32-S3

TD-er commented:
> > > Have you removed the shield metal from one of the modules? Maybe it is not a embedded flash? > Flash swap...
> > 
> > 
> > It's definitely an external flash version, I tried to swap it with 4MB chip, but no luck so far, it doesn't boot in serial download mode. Will try again
> 
> Thats for sure a solder issue.

Could it also be a compatibility issue like DIO/DOUT mode not supported by that specific chip, but needed due to wiring?

Jason2866 commented:
> > > > Have you removed the shield metal from one of the modules? Maybe it is not a embedded flash? > Flash swap...
> > > 
> > > 
> > > It's definitely an external flash version, I tried to swap it with 4MB chip, but no luck so far, it doesn't boot in serial download mode. Will try again
> > 
> > 
> > Thats for sure a solder issue.
> 
> Could it also be a compatibility issue like DIO/DOUT mode not supported by that specific chip, but needed due to wiring?

No, since it would fail with the 2MB flash too. The difference is just the flash size.

TD-er committed:
[Docs] Update documentation regarding supported ESP chips

tonhuisman commented on docs/source/ESPEasy/ESPchips.rst:
```diff
@@ -227,3 +296,17 @@ The only advantages of the ESP32-S2 compared to its predecessor are:
+
+Added: 2023/05/03
+
+.. note:: Support for the ESP32-S3 is very preliminary, as in it is hardly tested (as of May 2023)
```
This line should mention ESP-**C**3

TD-er committed:
[NeoPixelBus] Work-around to compile on older compilers (ESP8266 2.7.4)

TD-er committed:
[docs] Fix copy/paste error in ESP32-C3 docs

chemmex commented:
ESP32-C3
Tried several I2C sensors, serial sensors, ST7789 LCD - all work good so far

Thanks!

chemmex commented:
Deep sleep works well, wakeup takes 2 seconds with Use Last Connected AP from RTC vs 9 seconds without it.

TD-er commented:
Check!
So it seems at least the C3 is working really well :)

Haven't had the chance yet to test with hardware connected to the S3, except for the display on the LilyGo T-dongle S3.

TD-er committed:
[Docs] Update specs of supported (and not yet supported) ESP modules

TD-er committed:
Merge remote-tracking branch 'origin/feature/ESP32_C3' into feature/ESP32_C3

TD-er committed:
[Docs] Document future ESP32 C2/C6/H2

TD-er committed:
[Cleanup] Simplify detection of active flash mode

TD-er committed:
Merge branch 'mega' into feature/ESP32_C3

TD-er committed:
[p2p] Fix task value data exchange between nodes via p2p (#4649)

Fixes: #4649

TD-er committed:
[p2p] Fix check on received data for matching plugin IDs

TD-er committed:
[p2p] Fix collecting stats + send to controller on p2p received data

TD-er committed:
[p2p] Fix build on LIMIT_BUILD_SIZE builds

TD-er committed:
[p2p] Fix receiving p2p data on another node (stupid copy/paste bug)

TD-er committed:
[ESP32-S2/S3/C3]  Update user configurable SPI bus in ESPEasy.

TD-er committed:
Merge branch 'mega' into feature/ESP32_C3

TD-er committed:
[Build] Fix PlatformIO env definition.

TD-er committed:
[Build] Fix build for ESP8266 due to misplaced #ifdef ESP32

TD-er committed:
[ESP32-S3/C3] Generate web flasher manifest files for all ESP platforms

TD-er committed:
Merge branch 'mega' into feature/ESP32_C3

TD-er committed:
[WiFi] Make Wi-Fi TX power configurable for ESP32-S2/S3/C3

tonhuisman commented on src/src/Helpers/Hardware.cpp:
```diff
@@ -674,6 +690,22 @@ uint32_t getFlashChipSpeed() {
+      case FM_QOUT:    return F("QOUT (OPI Wired)");
+      case FM_DIO:     return F("DIO (OPI Wired)");
+      case FM_DOUT:    return F("DOUT (OPI Wired)");
+  #ifdef ESP32
```
This `#ifdef` is already active here

TD-er replied:
Yep, not sure if I keep this code as it seems impossible to detect whether flash or PSRAM is connected using 8-pin SPI.
The only way to be somewhat sure about this is when PSRAM isn't working when flashing using QIO/QSPI, but is found when using QIO/OPI mode.
But that's hardly "runtime" detection :(

TD-er committed:
[ESP32 SDK] Update to ESP32 SDK 2.0.9 + Enable WiFi TX pwr for ESP32

TD-er committed:
[PIO build] Update build scripts for scons 4.5.x

TD-er committed:
[Build] Fix build error due to strange cast of double type in rules

TD-er committed:
[Build] Work-around for new ESP8266 macro of round()

Whoever thought it would be a good idea to make round() an explicit cast to `long`....

TD-er committed:
[Build ESP8266] Replace calls to round()

TD-er committed:
[Build] Use HWrandom to overcome build issues

TD-er committed:
[Build] Fix missing include

TD-er committed:
[Build] Remove extern "C" to allow for (C++) function overloading

TD-er committed:
[Build] include ESPEasy_common.h instead of Arduino.h where possible

TD-er committed:
[Cleanup] Moving not commonly needed code out of ESPEasy_common.h

TD-er committed:
[Build] Remove "Helpers" folder from concat script

TD-er committed:
[Build] Include ESPEasy_common.h instead of ESPEasy_config.h

TD-er committed:
[Build] Add Helpers folder to concat .cpp for ESP8266

TD-er committed:
[Build] Add ESPEasyCore to concat cpp & split global RulesCalculate obj

TD-er committed:
[Build test] Disable lots of plugins + not concat Helpers/*.cpp

TD-er committed:
[Build test] Disable P037 to test build failures

TD-er committed:
[Build test] Undef round right after include <Arduino.h>

TD-er committed:
[Build] Revert disabling plugins + concat Helpers/*.cpp

TD-er committed:
[Build] Include cmath after Arduino.h to redefine round()

TD-er committed:
Merge branch 'mega' into feature/ESP32_C3
